File: C++/rsa.cpp
#include <iostream>
#include <cmath>

// Greatest Common Divisor
int gcd(int a, int b) {
    while (b != 0) {
        int tmp = b;
        b = a % b;
        a = tmp;
    }
    return a;
}

// Modular Exponentiation (base^exp % mod)
long long mod_exp(long long base, long long exp, long long mod) {
    long long result = 1;
    base = base % mod;

    while (exp > 0) {
        if (exp % 2 == 1)
            result = (result * base) % mod;

        exp = exp >> 1;
        base = (base * base) % mod;
    }
    return result;
}

// Extended Euclidean Algorithm to find modular inverse
int mod_inverse(int e, int phi) {
    int t = 0, newt = 1;
    int r = phi, newr = e;

    while (newr != 0) {
        int quotient = r / newr;
        int temp = newt;
        newt = t - quotient * newt;
        t = temp;

        temp = newr;
        newr = r - quotient * newr;
        r = temp;
    }

    if (r > 1) return -1; // e is not invertible
    if (t < 0) t += phi;
    return t;
}

// RSA Example
int main() {
    // Choose two primes (small for example only!)
    int p = 61;
    int q = 53;

    // Compute n and phi(n)
    int n = p * q;
    int phi = (p - 1) * (q - 1);

    // Choose e (public exponent)
    int e = 17;
    while (gcd(e, phi) != 1)
        ++e;

    // Compute d (private exponent)
    int d = mod_inverse(e, phi);
    if (d == -1) {
        std::cerr << "Modular inverse failed.\n";
        return 1;
    }

    std::cout << "Public key: (" << e << ", " << n << ")\n";
    std::cout << "Private key: (" << d << ", " << n << ")\n";

    // Message to encrypt
    int message = 65;
    std::cout << "Original Message: " << message << "\n";

    // Encrypt: c = m^e mod n
    long long ciphertext = mod_exp(message, e, n);
    std::cout << "Encrypted Message: " << ciphertext << "\n";

    // Decrypt: m = c^d mod n
    long long decrypted = mod_exp(ciphertext, d, n);
    std::cout << "Decrypted Message: " << decrypted << "\n";

    return 0;
}

File: Python/rsa.py
def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a

def mod_inverse(e, phi):
    d = 0
    x1, x2, y1 = 0, 1, 1
    temp_phi = phi
    while e > 0:
        temp1, temp2 = temp_phi // e, temp_phi - (temp_phi // e) * e
        temp_phi, e = e, temp2
        x, y = x2 - temp1 * x1, d - temp1 * y1
        x2, x1, d, y1 = x1, x, y1, y
    if temp_phi == 1:
        return d + phi

def encrypt(public_key, plaintext):
    key, n = public_key
    cipher = [(ord(char) ** key) % n for char in plaintext]
    return cipher

def decrypt(private_key, ciphertext):
    key, n = private_key
    plain = [chr((char ** key) % n) for char in ciphertext]
    return ''.join(plain)

p = 53
q = 59
n = p * q
phi = (p - 1) * (q - 1)
e = 3
d = mod_inverse(e, phi)

public_key = (e, n)
private_key = (d, n)

message = "HELLO"
encrypted_msg = encrypt(public_key, message)
print("Encrypted message:", encrypted_msg)

decrypted_msg = decrypt(private_key, encrypted_msg)
print("Decrypted message:", decrypted_msg)